// @nx/vite/src/utils/versions.js
"use strict";
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    edgeRuntimeVmVersion: function() {
        return edgeRuntimeVmVersion;
    },
    happyDomVersion: function() {
        return happyDomVersion;
    },
    jsdomVersion: function() {
        return jsdomVersion;
    },
    nxVersion: function() {
        return nxVersion;
    },
    vitePluginDtsVersion: function() {
        return vitePluginDtsVersion;
    },
    vitePluginReactSwcVersion: function() {
        return vitePluginReactSwcVersion;
    },
    vitePluginReactVersion: function() {
        return vitePluginReactVersion;
    },
    viteVersion: function() {
        return viteVersion;
    },
    vitestCoverageIstanbulVersion: function() {
        return vitestCoverageIstanbulVersion;
    },
    vitestCoverageV8Version: function() {
        return vitestCoverageV8Version;
    },
    vitestVersion: function() {
        return vitestVersion;
    }
});
const nxVersion = require('../../package.json').version;
const viteVersion = '~5.0.0';
const vitestVersion = '^1.3.1';
const vitePluginReactVersion = '^4.2.0';
const vitePluginReactSwcVersion = '^3.5.0';
const jsdomVersion = '~22.1.0';
const vitePluginDtsVersion = '~3.8.1';
const happyDomVersion = '~9.20.3';
const edgeRuntimeVmVersion = '~3.0.2';
const vitestCoverageV8Version = '^1.0.4';
const vitestCoverageIstanbulVersion = '^1.0.4';
