// @nx/esbuild/src/generators/configuration/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.configurationGenerator = void 0;
const devkit_1 = require("@nx/devkit");
const get_import_path_1 = require("@nx/js/src/utils/get-import-path");
const init_1 = require("../init/init");
const add_build_target_defaults_1 = require("@nx/devkit/src/generators/add-build-target-defaults");
async function configurationGenerator(tree, options) {
    const task = await (0, init_1.esbuildInitGenerator)(tree, {
        ...options,
        skipFormat: true,
    });
    options.buildTarget ??= 'build';
    checkForTargetConflicts(tree, options);
    addBuildTarget(tree, options);
    await (0, devkit_1.formatFiles)(tree);
    return task;
}
exports.configurationGenerator = configurationGenerator;
function checkForTargetConflicts(tree, options) {
    if (options.skipValidation)
        return;
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    if (project.targets?.[options.buildTarget]) {
        throw new Error(`Project "${options.project}" already has a ${options.buildTarget} target. Pass --skipValidation to ignore this error.`);
    }
}
function addBuildTarget(tree, options) {
    (0, add_build_target_defaults_1.addBuildTargetDefaults)(tree, '@nx/esbuild:esbuild', options.buildTarget);
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    const packageJsonPath = (0, devkit_1.joinPathFragments)(project.root, 'package.json');
    if (!tree.exists(packageJsonPath)) {
        const importPath = options.importPath || (0, get_import_path_1.getImportPath)(tree, options.project);
        (0, devkit_1.writeJson)(tree, packageJsonPath, {
            name: importPath,
            version: '0.0.1',
        });
    }
    const prevBuildOptions = project.targets?.[options.buildTarget]?.options;
    const tsConfig = prevBuildOptions?.tsConfig ?? getTsConfigFile(tree, options);
    const buildOptions = {
        main: prevBuildOptions?.main ?? getMainFile(tree, options),
        outputPath: prevBuildOptions?.outputPath ??
            (0, devkit_1.joinPathFragments)('dist', project.root === '.' ? options.project : project.root),
        outputFileName: 'main.js',
        tsConfig,
        assets: [],
        platform: options.platform,
    };
    if (options.platform === 'browser') {
        buildOptions.outputHashing = 'all';
        buildOptions.minify = true;
    }
    if (tree.exists((0, devkit_1.joinPathFragments)(project.root, 'README.md'))) {
        buildOptions.assets = [
            {
                glob: `${project.root}/README.md`,
                input: '.',
                output: '.',
            },
        ];
    }
    (0, devkit_1.updateProjectConfiguration)(tree, options.project, {
        ...project,
        targets: {
            ...project.targets,
            [options.buildTarget]: {
                executor: '@nx/esbuild:esbuild',
                outputs: ['{options.outputPath}'],
                defaultConfiguration: 'production',
                options: buildOptions,
                configurations: {
                    development: {
                        minify: false,
                    },
                    production: {
                        minify: true,
                    },
                },
            },
        },
    });
}
function getMainFile(tree, options) {
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    const candidates = [
        (0, devkit_1.joinPathFragments)(project.root, 'src/main.ts'),
        (0, devkit_1.joinPathFragments)(project.root, 'src/index.ts'),
    ];
    for (const file of candidates) {
        if (tree.exists(file))
            return file;
    }
    return options.main;
}
function getTsConfigFile(tree, options) {
    const project = (0, devkit_1.readProjectConfiguration)(tree, options.project);
    const candidates = [
        (0, devkit_1.joinPathFragments)(project.root, 'tsconfig.lib.json'),
        (0, devkit_1.joinPathFragments)(project.root, 'tsconfig.app.json'),
    ];
    for (const file of candidates) {
        if (tree.exists(file))
            return file;
    }
    return options.tsConfig;
}
exports.default = configurationGenerator;
